commands2/mecanumcontrollercommand.py

```diff
@@ -0,0 +1,216 @@
+    internally. This is intended to be a more-or-less "complete solution" that can be used by teams
+    without a great deal of controls expertise.
+
+    Advanced teams seeking more flexibility (for example, those who wish to use the onboard PID
+    functionality of a "smart" motor controller) may use the secondary constructor that omits the PID
+    and feedforward functionality, returning only the raw wheel speeds from the PID controllers.
+
+    The robot angle controller does not follow the angle given by the trajectory but rather goes
+    to the angle given in the final state of the trajectory.
+    """
+
+    def __init__(
+        self,
+        trajectory: Trajectory,
+        pose: Callable[[], Pose2d],
+        kinematics: MecanumDriveKinematics,
+        xController: PIDController,
+        yController: PIDController,
+        thetaController: ProfiledPIDControllerRadians,
+        # rotationSupplier: Callable[[], Rotation2d],
+        maxWheelVelocityMetersPerSecond: float,
+        outputConsumer: Callable[[MecanumDriveWheelSpeeds], None],
+        *requirements: Subsystem,
+        feedforward: Optional[SimpleMotorFeedforwardMeters] = None,
+        frontLeftController: Optional[PIDController] = None,
+        rearLeftController: Optional[PIDController] = None,
+        frontRightController: Optional[PIDController] = None,
+        rearRightController: Optional[PIDController] = None,
+        currentWheelSpeedsSupplier: Optional[
+            Callable[[], MecanumDriveWheelSpeeds]
+        ] = None,
+    ):
+        super().__init__()
+
+        self._trajectory: Trajectory = trajectory
+        self._pose = pose
+        self._kinematics = kinematics
+        self._controller = HolonomicDriveController(
+            xController, yController, thetaController
+        )
+        # self.rotationSupplier = rotationSupplier
+        self._maxWheelVelocityMetersPerSecond = maxWheelVelocityMetersPerSecond
+        self._outputConsumer = outputConsumer
+
+        # Member to track time of last loop execution
+        self._prevTime = 0
+        self._usePID = False
+
+        # Optional requirements checks.  If any one of the feedforward, pidcontrollers, or wheelspeedsuppliers
+        # are not None, then all should be non-null.  Raise RuntimeError if they are. The List of PID controllers
+        # should be equal to each corner of the robot being commanded.
+        if (
+            feedforward
+            or frontLeftController
+            or rearLeftController
+            or frontRightController
+            or rearRightController
+            or currentWheelSpeedsSupplier is not None
+        ):
+            if (
+                feedforward
+                or frontLeftController
+                or rearLeftController
+                or frontRightController
+                or rearRightController
+                or currentWheelSpeedsSupplier is None
+            ):
+                raise RuntimeError(
```

Comment:
If any of

- feedforward
- frontLeftController
- rearLeftController
- frontRightController
- rearRightController

are provided (i.e. not `None`), won't we always get here? This doesn't seem right.